### ISSUE TITLE
make the standalone version work on ARM (armhf)

### ISSUE THREAD

**Issue by fczuardi** [zedapp/zed#510]
I find it super cool that Zed is currently the best text editor available for ChromeOS, it is one of the top apps that can turn cheap toy-laptops such as $200~ chromebooks into serious development machines. 

Wouldn't it be cool if Zed was also able to run as a standalone editor for people with such machines running ubuntu, or some other linux distro (or their chroot jails with crouton)? Or eventually raspberryPi and other small computers with arm boards :)

This bug is about the tasks to make that happen, I kinda of hacked my way through into a bootable standalone zed version on a crouton-ubuntu-trusty with x11 and iojs, some of the things that have to be patched:

- [ ] update Makefile to detect if the linux 32 is armhf
- [ ] update Makefile to download the correct nw.js binaries for this architecture (we can use some old version of the compiled nw like https://github.com/toxygen/armhf-node-webkit while the proper fix for oficial releases is not there yet https://github.com/nwjs/nw.js/issues/1151)
  - [ ] or throw a message saying that the architecture is not yet supported if someone tries to "make" while this bug is open
- if we were to use https://github.com/toxygen/armhf-node-webkit
  - [ ] it would be nice to have a newer version of nw compiled( for example, the current one doesnt support Promises https://github.com/toxygen/armhf-node-webkit/issues/4 )
  - [ ] with the current old binary, polyfill whatever pieces of code that the old webkit does not support (so far it looks like just Promise and console globals)
- [ ] for ubuntu, include messages of how to fix/symlink the two not found libraries: [libudev.so.0](https://github.com/nwjs/nw.js/wiki/The-solution-of-lacking-libudev.so.0) and [nss](http://raspberrypi.stackexchange.com/a/13707)

**Comment by fczuardi** [comment]
Bug: Ctrl+P on markdown files opens a blank preview

**Comment by zefhemel** [comment]
This would be nice to have, of course! Of course, the chrome app would work already, but a standalone version -- why not?

FYI: I have an experimental branch nwjs (https://github.com/zedapp/zed/tree/nwjs) that contains some work to make Zed work with the "successor" of node-webkit nw.js (it's really the same thing, just using io.js and a new name). Perhaps it would sense to look into that? I don't know if they have ARM support (or plans for ARM support).

**Comment by fczuardi** [comment]
There is an open issue on the project, https://github.com/nwjs/nw.js/issues/1151 it is there since 2013 so I wouldn't hold my breath, but we can use the binaries for ARM generated elsewhere meanwhile

**Comment by fczuardi** [comment]
first task implemented on https://github.com/fczuardi/zed/commit/984721df6287d875c394c5f9265b24d9b9ec0880

**Comment by fczuardi** [comment]
with the version 0.12.0 of nw, we need to copy one extra file from the nw release: ```icudtl.dat```, see https://github.com/nwjs/nw.js/wiki/How-to-package-and-distribute-your-apps#which-files-should-be-shipped

**Comment by fczuardi** [comment]
with https://github.com/fczuardi/zed/commit/5f3f0dd27a93594adcb933dbfa8163f1cc45677b nwjs seems to work, the webkit window open with the zed icon in place, the remaining errors looks like application / javascript related, here is the log of the errors when zed-bin is launched:

```
[12813:0321/201717:WARNING:resource_bundle.cc(315)] locale_file_path.empty()
[12813:0321/201717:ERROR:browser_main_loop.cc(170)] Running without the SUID sandbox! See https://code.google.com/p/chromium/wiki/LinuxSUIDSandboxDevelopment for more information on developing with the sandbox on.
[12815:0321/201717:WARNING:resource_bundle.cc(315)] locale_file_path.empty()
[12813:0321/201717:INFO:audio_manager_pulse.cc(258)] Failed to connect to the context.  Error: Connection refused
[12832:0321/201718:WARNING:resource_bundle.cc(315)] locale_file_path.empty()
[12813:0321/201718:WARNING:resource_bundle.cc(512)] locale resources are not loaded
[12813:0321/201718:WARNING:resource_bundle.cc(512)] locale resources are not loaded
[12813:0321/201718:WARNING:resource_bundle.cc(512)] locale resources are not loaded
[12813:0321/201718:WARNING:resource_bundle.cc(512)] locale resources are not loaded
[12813:0321/201720:INFO:CONSOLE(114)] ""Architect resolved"", source: app://./js/boot.js (114)
[12813:0321/201720:INFO:CONSOLE(178)] ""Loaded eventbus"", source: app://./js/boot.js (178)
[12813:0321/201720:INFO:CONSOLE(178)] ""Loaded ui"", source: app://./js/boot.js (178)
[12813:0321/201720:INFO:CONSOLE(178)] ""Loaded keys"", source: app://./js/boot.js (178)
[12813:0321/201720:INFO:CONSOLE(178)] ""Loaded db"", source: app://./js/boot.js (178)
[12813:0321/201720:INFO:CONSOLE(505)] "Uncaught (in promise) TypeError: Cannot set property 'name' of null", source: app://./dep/architect.js (505)
```

**Comment by fczuardi** [comment]
I am investigating and I am unfamiliar with the codebase, but apparently it is something with the "background" plugin being null, does it ring any bells @zefhemel ?

**Comment by fczuardi** [comment]
ok, the first thing is that this variable https://github.com/zedapp/zed/blob/master/app/js/background.js#L7 window.isNodeWebkit is not there.

**Comment by fczuardi** [comment]
yeah! got it working, the check on https://github.com/zedapp/zed/blob/master/app/js/boot.js#L13 will need to change.

when I changed to ```window.isNodeWebkit = true``` just to check, everything started working :)

**Comment by fczuardi** [comment]
This are the properties of window on at that point on boot.js:

"top","window","location","nwDispatcher","document","root","global","process","Buffer","require","__nwWindowId","chrome","nodeRequire","_","goog","analytics","JSON5","requirejs","define","$","jQuery","_log","_checkBrowser","logMsg","_canLog","BROWSER","getDynaTreePersistData","DTNodeStatus_Error","DTNodeStatus_Loading","DTNodeStatus_Ok","foo","i","speechSynthesis","webkitStorageInfo","indexedDB","webkitIndexedDB","crypto","localStorage","sessionStorage","applicationCache","CSS","performance","console","devicePixelRatio","styleMedia","parent","opener","frames","self","defaultstatus","defaultStatus","status","name","length","closed","pageYOffset","pageXOffset","scrollY","scrollX","screenTop","screenLeft","screenY","screenX","innerWidth","innerHeight","outerWidth","outerHeight","frameElement","clientInformation","navigator","toolbar","statusbar","scrollbars","personalbar","menubar","locationbar","history","screen","ondeviceorientation","ondevicemotion","postMessage","close","blur","focus","onunload","onstorage","onpopstate","onpageshow","onpagehide","ononline","onoffline","onmessage","onlanguagechange","onhashchange","onbeforeunload","onwaiting","onvolumechange","ontoggle","ontimeupdate","onsuspend","onsubmit","onstalled","onshow","onselect","onseeking","onseeked","onscroll","onresize","onreset","onratechange","onprogress","onplaying","onplay","onpause","onmousewheel","onmouseup","onmouseover","onmouseout","onmousemove","onmouseleave","onmouseenter","onmousedown","onloadstart","onloadedmetadata","onloadeddata","onload","onkeyup","onkeypress","onkeydown","oninvalid","oninput","onfocus","onerror","onended","onemptied","ondurationchange","ondrop","ondragstart","ondragover","ondragleave","ondragenter","ondragend","ondrag","ondblclick","oncuechange","oncontextmenu","onclose","onclick","onchange","oncanplaythrough","oncanplay","oncancel","onblur","onabort","onwheel","onwebkittransitionend","onwebkitanimationstart","onwebkitanimationiteration","onwebkitanimationend","ontransitionend","onsearch","getSelection","print","stop","open","alert","confirm","prompt","find","moveBy","moveTo","resizeBy","resizeTo","matchMedia","getComputedStyle","getMatchedCSSRules","requestAnimationFrame","cancelAnimationFrame","webkitRequestAnimationFrame","webkitCancelAnimationFrame","webkitCancelRequestAnimationFrame","captureEvents","releaseEvents","btoa","atob","setTimeout","clearTimeout","setInterval","clearInterval","scrollBy","scrollTo","scroll","TEMPORARY","PERSISTENT","webkitRequestFileSystem","webkitResolveLocalFileSystemURL","openDatabase","addEventListener","removeEventListener","dispatchEvent"

So chrome is not undefined, we will need to use something else :)

**Comment by fczuardi** [comment]
maybe nwDispatcher?

```javascript
window.isNodeWebkit = window.nwDispatcher !== undefined;
```

**Comment by fczuardi** [comment]
or process?

```javascript
window.isNodeWebkit = window.process !== undefined;
```

**Comment by fczuardi** [comment]
With https://github.com/fczuardi/zed/commit/e7f8366f1b5909ed0d5113c56530294db1e3d12d I got it running, should I squash the 3 changes in a pull request?

**Comment by fczuardi** [comment]
Oh, I see that the branch you have suggested already included the changes for icudtl.dat and isNodeWebkit, so the only thing left is the Makefile tweak to include conditionals for arm processor (uname -p) and the download url for the temporary unnoficial/unsupported binary release of nwjs at https://github.com/jtg-gg/node-webkit/releases/tag/nw-v0.12.0
